osquery/tables/system/linux/md_tables.cpp

```diff
@@ -0,0 +1,771 @@
+ * extracted
+ *
+ * @return bool indicating whether a line for the searchTerm was found in the
+ * line n + 1 and handleFunc was executed, increments param n to n + 1
+ */
+static inline bool handleMDStatuses(
+    const std::vector<std::string>& lines,
+    size_t& n,
+    const std::string& searchTerm,
+    std::function<void(const std::string& line)> handleFunc) {
+  std::size_t pos = lines[n + 1].find(searchTerm);
+  if (pos != std::string::npos) {
+    std::string line(lines[n + 1].substr(pos + searchTerm.length()));
+    boost::algorithm::trim(line);
+    handleFunc(line);
+    // Add an extra line for next iteration if so..
+    n += 1;
+    return true;
+  }
+
+  return false;
+}
+
+MDDrive parseMDDrive(const std::string& name) {
+  MDDrive drive;
+  drive.name = name;
+
+  auto start = name.find('[');
+  auto end = name.find(']');
+  if (start == std::string::npos || end == std::string::npos) {
+    LOG(WARNING) << "Unexpected drive name format: " << name;
+    return drive;
+  }
+
+  drive.pos = std::stoi(name.substr(start + 1, end - start - 1));
+
+  return drive;
+}
+
+void MD::parseMDStat(const std::vector<std::string>& lines, MDStat& result) {
+  // Will be used to determine starting point of lines to work on.
+  size_t n = 0;
+
+  if (lines.empty()) {
+    return;
+  }
+
+  // This should always evaluate to true, but just in case we check.
+  if (lines[0].find("Personalities :") != std::string::npos) {
+    std::string pline(lines[0].substr(sizeof("Personalities :") - 1));
+    parseMDPersonalities(pline, result.personalities);
+
+    n = 1;
+
+  } else {
+    LOG(WARNING) << "mdstat Personalites not found at line 0: " << lines[0];
+  }
+
+  while (n < lines.size()) {
+    if (lines[n].find_first_not_of("\t\r\v ") == std::string::npos) {
+      n += 1;
+      continue;
+    }
+    // Work off of first 2 character instead of just the first to be safe.
+    std::string firstTwo = lines[n].substr(0, 2);
+    if (firstTwo == "md") {
+      auto mdline(split(lines[n], ":", 1));
+      if (mdline.size() < 2) {
+        LOG(WARNING) << "Unexpected md device line structure: " << lines[n];
+        continue;
+      }
+
+      MDDevice mdd;
+      mdd.name = std::move(mdline[0]);
+      boost::algorithm::trim(mdd.name);
+
+      auto settings(split(mdline[1], " "));
+      trimStrs(settings);
+      // First 2 of settings are always status and RAID level
+      if (settings.size() >= 2) {
+        mdd.status = std::move(settings[0]);
+        mdd.raidLevel = std::move(settings[1]);
+
+        for (size_t i = 2; i < settings.size(); i++) {
+          mdd.drives.push_back(parseMDDrive(settings[i]));
+        }
+      }
+
+      /* Next line is device config and settings.  We handle here instead of
+       * later b/c pieces are need for both md_drives and md_devices table.  For
+       * safety, we check if we at the end of the file. */
+      if (n >= lines.size() - 1) {
```

Comment:
I used  `n >= lines.size() -1` for safety issues, i.e. just in case the code above (improperly) increments n to `> lines.size()` , we don't just proceed forward down the block.